File: Week_08/146/lru_cache.go
package main

type DNode struct {
	key   int
	value int
	prev  *DNode
	next  *DNode
}

type LRUCache struct {
	cap   int
	root  *DNode
	cache map[int]*DNode
}

func Constructor(capacity int) LRUCache {
	root := &DNode{}
	root.prev = root
	root.next = root
	return LRUCache{
		cap:   capacity,
		root:  root,
		cache: map[int]*DNode{},
	}
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.cache[key]; ok {
		this.movetoEnd(node)
		return node.value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.cache[key]; ok {
		node.value = value
		this.movetoEnd(node)
	} else {
		if len(this.cache) < this.cap {
			node := &DNode{
				key:   key,
				value: value,
				prev:  this.root.prev,
				next:  this.root,
			}
			this.root.prev.next = node
			this.root.prev = node
			this.cache[key] = node
		} else {
			this.root.key, this.root.value = key, value
			this.cache[key] = this.root
			this.root = this.root.next
			delete(this.cache, this.root.key)
		}
	}
}

func (this *LRUCache) movetoEnd(node *DNode) {
	node.prev.next, node.next.prev = node.next, node.prev
	node.prev, node.next = this.root.prev, this.root
	this.root.prev.next, this.root.prev = node, node
}

File: Week_08/146/lru_cache.py
class DNode:
    __slots__ = ("key", "value", "prev", "next")


class LRUCache(object):
    def __init__(self, capacity):
        self.cap = capacity
        self.root = DNode()
        self.root.next, self.root.prev = self.root, self.root
        self.root.key = self.root.value = -1
        self.cache = {}

    def move_to_end(self, node: DNode):
        node.next.prev, node.prev.next = node.prev, node.next
        node.prev, node.next = self.root.prev, self.root
        self.root.prev.next, self.root.prev = node, node

    def get(self, key: int):
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.move_to_end(node)
        return node.value

    def put(self, key: int, value: int):
        if key in self.cache:
            self.cache[key].value = value
            self.move_to_end(self.cache[key])
        else:
            if len(self.cache) < self.cap:
                node = DNode()
                node.key, node.value = key, value
                node.prev, node.next = self.root.prev, self.root
                self.root.prev.next, self.root.prev = node, node
                self.cache[key] = node
            else:
                self.root.key = key
                self.root.value = value
                self.cache[key] = self.root
                self.cache.pop(self.root.next.key)
                self.root = self.root.next
